### PR TITLE
feat: support Peer ID represented as CID

### PR DESCRIPTION
**Pull request by lidel**
### Motivation

This PR implements `Stage 1: Parse CIDs as peer ID` from https://github.com/libp2p/specs/issues/216. 

**TL;DR** This PR does not change the default text representation of Peer IDs, all we do in `Stage 1` is adding support for creating PeerId objects from CIDs.  

For a wider context see [libp2p/specs/RFC/0001-text-peerid-cid.md](https://github.com/libp2p/specs/blob/master/RFC/0001-text-peerid-cid.md), short story is that CID support will enable IPFS project to support `/ipns/{cid}` in Base32 (https://github.com/ipfs/ipfs/issues/337), enabling things like `http://{libp2p-key-as-cidv1b32}.ipns.dweb.link`

### Changes

This PR adds two functions:

- `createFromCID` accepts CID as String|CID|Buffer and creates PeerId from the multihash value inside of it
- `toCIDString` serializes PeerId multihash into a CIDv1 in Base32 ([libp2p RFC 0001](https://github.com/libp2p/specs/blob/master/RFC/0001-text-peerid-cid.md))

### Next

- We need this released and bubbled up, so we can switch IPNS implementation in js-ipfs to `createFromCID`  in https://github.com/ipfs/js-ipfs/pull/2566

cc https://github.com/ipfs/ipfs/issues/337, https://github.com/libp2p/specs/issues/216, https://github.com/libp2p/specs/pull/209, https://github.com/ipfs/go-ipfs/issues/5287

### CODE REVIEW

File: src/index.js [hunk 1]
```diff
@@ -122,6 +123,12 @@ class PeerId {
     return this._idB58String
   }
 
+  // return string representation from RFC 0001: https://github.com/libp2p/specs/pull/209
+  toCIDString () {
+    const cid = new CID(1, 'libp2p-key', this.id, 'base32')
+    return cid.toBaseEncodedString('base32')
```
Comment (lidel) [inline src/index.js]: Given this is `Stage 1`, I was not sure if we should generate CIDString in constructor, or on-demand every time `toCIDString` is called.  Let me know if I should replace on-demand generation with  `return this._idCIDString`.

File: src/index.js [hunk 2]
```diff
@@ -187,6 +194,18 @@ exports.createFromB58String = (str) => {
   return new PeerIdWithIs(mh.fromB58String(str))
```
Comment (lidel) [inline src/index.js]: Given the fact that B58String is CIDv0, should we replace this with below? 
It would add support for CID, even if someone forgot to replace  `createFromB58String` with `createFromCID`:

```suggestion
  return exports.createFromCID(str)
```

File: src/index.js (hunk 1 above)
Comment (jacobheun) [reply src/index.js]: Lazy generation would be good. I know it gets called a lot, especially for debug logs so not having to compute that every time would be great.

Also, what do you think about just making this `toString`? We could potentially overload the function to support other types, but I think it would be great to get away from using `to<format>String`. The goal is to use cids, so moving towards that being the default of `toString` feels like the right thing to do to me.

File: src/index.js (hunk 2 above)
Comment (jacobheun) [reply src/index.js]: This seems reasonable to me

File: src/index.js (hunk 1 above)
Comment (lidel) [reply src/index.js]: Makes sense :+1: CIDv1 is self-describing (has multicodec which says "this is libp2p-key").
Added lazy-generation with memory and renamed to `toString` in https://github.com/libp2p/js-peer-id/pull/105/commits/60218d021a599239db2a10686be84033045b50fc

Comment (daviddias) [reply src/index.js]: A multicodec for libp2p-keys is a thing now? I need to catch up!

File: README.md [hunk 3]
```diff
@@ -197,9 +207,18 @@ Returns the Peer ID's `id` as a buffer.
 <Buffer 12 20 d6 24 39 98 f2 fc 56 34 3a d7 ed 03 42 ab 78 86 a4 eb 18 d7 36 f1 b6 7d 44 b3 7f cc 81 e0 f3 9f>
 ```
 
+
+### `toString()`
```
Comment (alanshaw) [inline README.md]: Personally I would add `toCID` and have `toString` as `() => this.toCID().toString()`.

This creates a symmetric API `createFromCID`/`toCID` (although "create" is superfluous IMHO, but meh, maybe a refactor for another day) and a nice sensible representation as a string that also enables `${peerId}` or `peerID + ''` without having to explicitly call `toString`.

File: test/peer-id.spec.js [hunk 4]
```diff
@@ -63,27 +66,68 @@ describe('PeerId', () => {
...
+  })
+
+  it('recreate from CID Buffer', () => {
+    const id = PeerId.createFromCID(testIdCID.buffer)
+    expect(testIdCIDString).to.equal(id.toString())
+    expect(testIdBytes).to.deep.equal(id.toBytes())
+  })
+
+  it('throws on invalid CID value', () => {
+    const invalidCID = 'QmaozNR7DZHQK1ZcU9p7QdrshMvXqWK6gpu5rmrkPdT3L'
```
Comment (jacobheun) [inline test/peer-id.spec.js]: Could we add a comment on why this is an invalid cid for future reference? The error doesn't make it terribly clear and at first glance it looks like a valid b58 id.

File: src/index.js [hunk 5]
```diff
@@ -122,6 +123,16 @@ class PeerId {
     return this._idB58String
```
Comment (alanshaw) [inline src/index.js]: This will now become rarely used. Should we generate on demand instead of in the constructor?

File: src/index.js [hunk 6]
```diff
@@ -122,6 +123,16 @@ class PeerId {
...
   }
 
+  // return self-describing String representation
+  // in default format from RFC 0001: https://github.com/libp2p/specs/pull/209
+  toString () {
+    if (!this._idCIDString) {
+      const cid = new CID(1, 'libp2p-key', this.id, 'base32')
+      this._idCIDString = cid.toBaseEncodedString('base32')
```
Comment (alanshaw) [inline src/index.js]: If you cache the CID instance you could use it in a `toCID` also. Note that CID's cache their string representation so calling `toString` on it won't re-encode.

File: src/index.js [hunk 7]
```diff
@@ -184,7 +195,19 @@ exports.createFromBytes = (buf) => {
 }
 
 exports.createFromB58String = (str) => {
-  return new PeerIdWithIs(mh.fromB58String(str))
+  return exports.createFromCID(str) // B58String is CIDv0
+}
+
+exports.createFromCID = (cid) => {
+  if (typeof cid === 'string' || Buffer.isBuffer(cid)) {
```
Comment (alanshaw) [inline src/index.js]: These checks are done in the CID constructor, we shouldn't repeat them here!

File: src/index.js [hunk 8]
```diff
@@ -184,7 +195,19 @@ exports.createFromBytes = (buf) => {
...
+    cid = new CID(cid)
+  } else if (CID.isCID(cid)) {
+    CID.validateCID(cid) // throws on error
```
Comment (alanshaw) [inline src/index.js]: If `CID.isCID` has returned true we _should_ have a valid CID instance. I don't think we need to validate it again do we?!

File: src/index.js [hunk 9]
```diff
@@ -184,7 +195,19 @@ exports.createFromBytes = (buf) => {
...
+  } else {
+    // provide more meaningful error than the one in CID.validateCID
+    throw new Error('Supplied cid value is neither String|CID|Buffer')
+  }
+  return new PeerIdWithIs(cid.multihash)
 }
```
Comment (alanshaw) [inline src/index.js]: I think this could be as simple as:

```js
exports.createFromCID = cid => {
  cid = CID.isCID(cid) ? cid : new CID(cid)
  if (cid.version === 1 && cid.codec !== 'libp2p-key') throw new Error('invalid peer ID')
  return new PeerIdWithIs(cid.multihash)
}
```

File: src/index.js [hunk 10]
```diff
@@ -184,7 +195,19 @@ exports.createFromBytes = (buf) => {
...
+
+exports.createFromCID = (cid) => {
+  if (typeof cid === 'string' || Buffer.isBuffer(cid)) {
+    cid = new CID(cid)
+  } else if (CID.isCID(cid)) {
+    CID.validateCID(cid) // throws on error
+  } else {
+    // provide more meaningful error than the one in CID.validateCID
+    throw new Error('Supplied cid value is neither String|CID|Buffer')
+  }
```
Comment (alanshaw) [inline src/index.js]: TODO: validate the codec is `libp2p-key` or that it's a v0?

File: src/index.js (hunk 1 above)
Comment (lidel) [reply src/index.js]: @daviddias https://github.com/multiformats/multicodec/issues/130 :ok_hand:

File: test/peer-id.spec.js (hunk 4 above)
Comment (lidel) [reply test/peer-id.spec.js]: The function code is `0x50`, which does not represent a hash function.
(we explicitly whitelist codes in [js-multihash/src/constants.js#L345](https://github.com/multiformats/js-multihash/blob/b85999d5768bf06f1b0f16b926ef2cb6d9c14265/src/constants.js#L345))
Will add this as a comment.

File: README.md (hunk 3 above)
Comment (lidel) [reply README.md]: I think there is no `toCID` because there was not `toMultihash`. 
We wanted to keep API surface small, liberal in inputs (accept CID object), but strict in outputs (just basic types: `PeerID`, `String` and `Buffer`)

File: src/index.js (hunk 5 above)
Comment (lidel) [reply src/index.js]: It is used in `isPeerId` and changing this triggers a bigger refactor.
I'd like to keep it as-is in this PR and change it in `Stage 2` of https://github.com/libp2p/specs/issues/216
(when we "Format peer IDs as CIDs by default.")

File: src/index.js (hunk 6 above)
Comment (lidel) [reply src/index.js]: As noted above, we don't want to expose CID types in outputs, just Strings.

File: src/index.js (hunk 8 above)
Comment (lidel) [reply src/index.js]: iiuc `isCID` comes from [`class-is`](https://www.npmjs.com/package/class-is) and only checks if `symbolName` is `@ipld/js-cid/CID`

File: src/index.js (hunk 7 above)
Comment (lidel) [reply src/index.js]: The reason for this was to provide a meaningful error `Supplied cid value is neither String|CID|Buffer`. 

If we delegate validation to CID constructor, it will spit out unhelpful  [`Invalid version, must be a number equal to 1 or 0`](https://github.com/multiformats/js-cid/blob/6c215bdc1a2440d5bccc0801327c1ef123e93a10/src/cid-util.js#L20) in most of error cases.

File: src/index.js (hunk 9 above)
Comment (lidel) [reply src/index.js]: Yeah I like this more :raised_hands:

File: src/index.js (hunk 10 above)
Comment (lidel) [reply src/index.js]: This is tricky. 
We want people to be able to convert CIDv0→CIDv1 and be able to use it.

See scenario below:

- Execute `ipfs id`
  - You will see PeerID as Base58 (CIDv0) which has [implicit multicodec `dag-pb`](https://github.com/multiformats/cid#cidv0)
- How to convert PeerID to CIDv1?
  - Probably via `ipfs cid base32` or https://cid.ipfs.io util
- Caveat: conversion produces a valid CIDv1 with (now explicit) `dag-pb` multicodec
  - the multihash inside of CIDv1 still represents a valid PeerID, and this library does not care about the rest od CID anyway

I think we should check for both `libp2p-key` and `dag-pb`. 
Will update the PR to do just that + tests.

**Comment by lidel** [comment]
ping @vasco-santos

File: README.md (hunk 3 above)
Comment (vasco-santos) [reply README.md]: I agree in keeping this in the context of this PR. We can iterate on this in a new PR if it justifies

**Comment by vasco-santos** [comment]
@alanshaw any other feedback that you want to provide here? I think we can go with this
